### PR TITLE
Several Commenter improvements

### PR DESCRIPTION
**Pull request by Turbo87**

### CODE REVIEW

**Comment by matklad** [comment]
This now depends on the PSI structure, which is far from solid at the moment. It may be worth to add a test =) But I don't know how to write a test for the thing so it may be too complicated or covered by some other kind of test.

**Comment by matklad** [comment]
Sorry, it actually depends only on the lexical structure, so a special test might be an overkill. But if it is easy to add, don't hesitate to do it =)

**Comment by Turbo87** [comment]
The problem is usually not writing the test, but finding out how to test it within the framework :wink:

**Comment by matklad** [comment]
Yep, that's why I love placeholder tests for stupid functionality: after hacking together a really cool and complicated feature it's really frustrating to figure out how to actually test it, so test are omitted.

**Comment by Turbo87** [comment]
:+1: my thoughts exactly

**Comment by matklad** [comment]
Found some similar tests in erlang plugin:

https://github.com/ignatov/intellij-erlang/blob/5f59ac290ca883c9f410f1d5eb672b7a19798403/tests/org/intellij/erlang/typing/ErlangEnterHandlerTest.java
https://github.com/ignatov/intellij-erlang/blob/5f59ac290ca883c9f410f1d5eb672b7a19798403/testData/autotemplates/enterInFunctionDocComment.erl
https://github.com/ignatov/intellij-erlang/blob/5f59ac290ca883c9f410f1d5eb672b7a19798403/testData/autotemplates/enterInFunctionDocComment-after.erl

**Comment by Turbo87** [comment]
@matklad awesome, I'll see what I can do. My goal was/is to automatically add the `///` on the next line if the current line has it. But since those are not *block* comments I don't think IntelliJ is able to do that automatically with the current approach. Might need some smart enter handling or something like that.

**Comment by Turbo87** [comment]
@matklad I've updated this PR with a few tests for the `Commenter` implementation

**Comment by Turbo87** [comment]
hrrr, it works :smiley: 

It is much more pleasant to write documentation comments now because you don't have to manually prefix them anymore. Once you've written the first line prefixed with `///` or `//!` the editor will automatically add that prefix again if you press "Enter".

File: src/org/rust/lang/typing/RustEnterInLineCommentHandler.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        val caret = CharArrayUtil.shiftForward(text, caretOffset.get(), " \t")
+
+        // figure out if the caret is at the end of the line
+        val isEOL = caret < text.length && text[caret] == '\n'
+
+        // find the PsiElement at the caret
+        var elementAtCaret = file.findElementAt(caret) ?: return Result.Continue
+        if (isEOL)
+            // ... or the previous one if this is end-of-line whitespace
+            elementAtCaret = elementAtCaret.prevSibling
```
Comment (atsky) [inline src/org/rust/lang/typing/RustEnterInLineCommentHandler.kt]: Please, use braces even for single statement if.

Comment (Turbo87) [inline src/org/rust/lang/typing/RustEnterInLineCommentHandler.kt]: done

**Comment by Turbo87** [comment]
awesome, thanks!

**Pull request merged**